yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+import { type IndexedTreeId, MerkleTreeId, type MerkleTreeReadOperations, getTreeHeight } from '@aztec/circuit-types';
+import { NullifierLeafPreimage, PublicDataTreeLeafPreimage } from '@aztec/circuits.js';
+import { poseidon2Hash } from '@aztec/foundation/crypto';
+import { Fr } from '@aztec/foundation/fields';
+import { type IndexedTreeLeafPreimage, type TreeLeafPreimage } from '@aztec/foundation/trees';
+
+import cloneDeep from 'lodash.clonedeep';
+
+/****************************************************/
+/****** Structs Used by the AvmEphemeralForest **/
+/****************************************************/
+
+/**
+ * The preimage and the leaf index of the Low Leaf (Low Nullifier or Low Public Data Leaf)
+ */
+export type LowPreimageWitness<T extends IndexedTreeLeafPreimage> = {
+  preimage: T;
+  index: bigint;
+};
+
+/**
+ * Contains the low sibling path and a boolean if the next index pointed to
+ * by the low leaf is an update or an insertion (i.e. exists or not).
+ */
+export type LowWitness<T extends IndexedTreeLeafPreimage> = {
+  preimage: T;
+  index: bigint;
+  update: boolean;
+  lowSiblingPath: Fr[];
+};
+
+/**
+ * The result of an indexed insertion in an indexed merkle tree.
+ * This will be used to hint the circuit
+ */
+export type IndexedInsertionResult<T extends IndexedTreeLeafPreimage> = {
+  leafIndex: bigint;
+  insertionPath: Fr[];
+  newOrElementToUpdate: { update: boolean; element: T };
+  lowWitness: LowWitness<T>;
+};
+
+/****************************************************/
+/****** The AvmEphemeralForest Class ****************/
+/****************************************************/
+
+/**
+ * This provides a forkable abstraction over the EphemeralAvmTree class
+ *  It contains the logic to look up into a read-only MerkleTreeDb to discover
+ *  the sibling paths and low witnesses that weren't inserted as part of this tx
+ */
+export class AvmEphemeralForest {
+  constructor(
+    public treeDb: MerkleTreeReadOperations,
+    public treeMap: Map<MerkleTreeId, EphemeralAvmTree>,
+    // This contains the preimage (as a buffer) and the leaf index of leaf in the ephemeral tree that contains the lowest "value" (i.e. nullifier value or public data tree slot)
+    public indexedTreeMin: Map<IndexedTreeId, [Buffer, bigint]>,
+    // This contains the indexed leaf preimages that were updated or inserted in the ephemeral tree and is keyed by leaf index
+    // This is needed since we have a sparse collectio of "value" sorted leaves in the ephemeral tree
+    public indexedUpdates: Map<IndexedTreeId, Map<bigint, Buffer>>,
```

Comment:
Hmm kinda - it will contain all the leaf changes to the indexed tree (although i didnt envision them being used as pending updates for the indexed trees). This is the linked list representation of the key-sorted indexed tree preimages. In a typical implementation you would represent this as a vector, but it's sparse for us so it's in a map for now.

This doesnt exist for the note hash tree - since we dont "sort" by keys, so I wouldn't make the tree class aware of it